File: im-mobile/lib/stores/self_destruct_message_store.dart
import 'dart:async';
import 'package:mobx/mobx.dart';
import '../api/self_destruct_message_api.dart';
import '../models/self_destruct_message_model.dart';

part 'self_destruct_message_store.g.dart';

/// 阅后即焚消息状态管理
/// 
/// @author IM Development Team
/// @since 1.0.0
class SelfDestructMessageStore = _SelfDestructMessageStore with _$SelfDestructMessageStore;

abstract class _SelfDestructMessageStore with Store {
  final SelfDestructMessageApi _api = SelfDestructMessageApi();

  // ========== 可观察状态 ==========

  @observable
  ObservableList<SelfDestructMessageModel> messages = ObservableList<SelfDestructMessageModel>();

  @observable
  ObservableList<SelfDestructMessageModel> sentMessages = ObservableList<SelfDestructMessageModel>();

  @observable
  ObservableList<SelfDestructMessageModel> receivedMessages = ObservableList<SelfDestructMessageModel>();

  @observable
  ObservableList<SelfDestructMessageModel> screenshotDetectedMessages = ObservableList<SelfDestructMessageModel>();

  @observable
  SelfDestructMessageModel? currentMessage;

  @observable
  SelfDestructConfig? config;

  @observable
  int unreadCount = 0;

  @observable
  bool isLoading = false;

  @observable
  String? errorMessage;

  @observable
  bool isReading = false;

  @observable
  int? countdownSeconds;

  // ========== 倒计时器 ==========

  Timer? _countdownTimer;

  // ========== 计算属性 ==========

  @computed
  List<SelfDestructMessageModel> get activeMessages => messages.where((m) => !m.isDestroyed).toList();

  @computed
  List<SelfDestructMessageModel> get destroyedMessages => messages.where((m) => m.isDestroyed).toList();

  @computed
  bool get hasUnread => unreadCount > 0;

  @computed
  bool get hasScreenshotDetected => screenshotDetectedMessages.isNotEmpty;

  // ========== 操作方法 ==========

  /// 获取配置
  @action
  Future<void> fetchConfig() async {
    try {
      isLoading = true;
      errorMessage = null;
      config = await _api.getConfig();
    } catch (e) {
      errorMessage = '获取配置失败: $e';
    } finally {
      isLoading = false;
    }
  }

  /// 获取会话消息
  @action
  Future<void> fetchMessages(String conversationId, {int page = 0, int size = 20}) async {
    try {
      isLoading = true;
      errorMessage = null;
      final result = await _api.getMessagesByConversation(conversationId, page: page, size: size);
      if (page == 0) {
        messages.clear();
      }
      messages.addAll(result);
    } catch (e) {
      errorMessage = '获取消息失败: $e';
    } finally {
      isLoading = false;
    }
  }

  /// 获取发送的消息
  @action
  Future<void> fetchSentMessages() async {
    try {
      isLoading = true;
      errorMessage = null;
      final result = await _api.getSentMessages();
      sentMessages.clear();
      sentMessages.addAll(result);
    } catch (e) {
      errorMessage = '获取发送消息失败: $e';
    } finally {
      isLoading = false;
    }
  }

  /// 获取接收的消息
  @action
  Future<void> fetchReceivedMessages() async {
    try {
      isLoading = true;
      errorMessage = null;
      final result = await _api.getReceivedMessages();
      receivedMessages.clear();
      receivedMessages.addAll(result);
    } catch (e) {
      errorMessage = '获取接收消息失败: $e';
    } finally {
      isLoading = false;
    }
  }

  /// 获取未读数量
  @action
  Future<void> fetchUnreadCount() async {
    try {
      final count = await _api.getUnreadCount();
      unreadCount = count;
    } catch (e) {
      errorMessage = '获取未读数量失败: $e';
    }
  }

  /// 获取会话未读数量
  @action
  Future<int> fetchUnreadCountByConversation(String conversationId) async {
    try {
      return await _api.getUnreadCountByConversation(conversationId);
    } catch (e) {
      errorMessage = '获取未读数量失败: $e';
      return 0;
    }
  }

  /// 创建消息
  @action
  Future<SelfDestructMessageModel?> createMessage(SelfDestructCreateRequest request) async {
    try {
      isLoading = true;
      errorMessage = null;
      final message = await _api.createMessage(request);
      messages.insert(0, message);
      sentMessages.insert(0, message);
      return message;
    } catch (e) {
      errorMessage = '创建消息失败: $e';
      return null;
    } finally {
      isLoading = false;
    }
  }

  /// 阅读消息
  @action
  Future<SelfDestructReadResponse?> readMessage(String messageId) async {
    try {
      isReading = true;
      errorMessage = null;
      final response = await _api.readMessage(messageId);
      
      // 更新消息状态
      final index = messages.indexWhere((m) => m.id == messageId);
      if (index != -1) {
        messages[index].markAsRead();
        messages[index].updateRemainingSeconds(response.remainingSeconds);
      }

      // 开始倒计时
      countdownSeconds = response.remainingSeconds;
      _startCountdown(messageId);

      return response;
    } catch (e) {
      errorMessage = '阅读消息失败: $e';
      return null;
    } finally {
      isReading = false;
    }
  }

  /// 开始倒计时
  void _startCountdown(String messageId) {
    _countdownTimer?.cancel();
    _countdownTimer = Timer.periodic(const Duration(seconds: 1), (timer) {
      if (countdownSeconds == null || countdownSeconds! <= 0) {
        timer.cancel();
        _destroyMessage(messageId);
      } else {
        countdownSeconds = countdownSeconds! - 1;
        final index = messages.indexWhere((m) => m.id == messageId);
        if (index != -1) {
          messages[index].updateRemainingSeconds(countdownSeconds!);
        }
      }
    });
  }

  /// 销毁消息
  void _destroyMessage(String messageId) {
    final index = messages.indexWhere((m) => m.id == messageId);
    if (index != -1) {
      messages[index].markAsDestroyed();
    }
    _api.destroyMessage(messageId).catchError((e) {
      // 忽略错误，可能已经被销毁
    });
  }

  /// 手动销毁消息
  @action
  Future<bool> destroyMessage(String messageId) async {
    try {
      await _api.destroyMessage(messageId);
      final index = messages.indexWhere((m) => m.id == messageId);
      if (index != -1) {
        messages[index].markAsDestroyed();
      }
      _countdownTimer?.cancel();
      return true;
    } catch (e) {
      errorMessage = '销毁消息失败: $e';
      return false;
    }
  }

  /// 删除消息
  @action
  Future<bool> deleteMessage(String messageId) async {
    try {
      await _api.deleteMessage(messageId);
      messages.removeWhere((m) => m.id == messageId);
      sentMessages.removeWhere((m) => m.id == messageId);
      return true;
    } catch (e) {
      errorMessage = '删除消息失败: $e';
      return false;
    }
  }

  /// 检测截图
  @action
  Future<ScreenshotDetectResponse?> detectScreenshot(String messageId) async {
    try {
      final response = await _api.detectScreenshot(messageId);
      final index = messages.indexWhere((m) => m.id == messageId);
      if (index != -1) {
        messages[index].recordScreenshot();
      }
      return response;
    } catch (e) {
      errorMessage = '截图检测失败: $e';
      return null;
    }
  }

  /// 获取被截图的消息列表
  @action
  Future<void> fetchScreenshotDetectedMessages() async {
    try {
      isLoading = true;
      errorMessage = null;
      final result = await _api.getScreenshotDetectedMessages();
      screenshotDetectedMessages.clear();
      screenshotDetectedMessages.addAll(result);
    } catch (e) {
      errorMessage = '获取截图检测消息失败: $e';
    } finally {
      isLoading = false;
    }
  }

  /// 检查消息是否已销毁
  @action
  Future<bool> isMessageDestroyed(String messageId) async {
    try {
      return await _api.isMessageDestroyed(messageId);
    } catch (e) {
      return true;
    }
  }

  /// 获取剩余秒数
  @action
  Future<int> getRemainingSeconds(String messageId) async {
    try {
      return await _api.getRemainingSeconds(messageId);
    } catch (e) {
      return 0;
    }
  }

  /// 添加新消息（从WebSocket接收）
  @action
  void addNewMessage(SelfDestructMessageModel message) {
    if (!messages.any((m) => m.id == message.id)) {
      messages.insert(0, message);
      receivedMessages.insert(0, message);
      unreadCount++;
    }
  }

  /// 更新消息（从WebSocket接收）
  @action
  void updateMessage(SelfDestructMessageModel message) {
    final index = messages.indexWhere((m) => m.id == message.id);
    if (index != -1) {
      messages[index] = message;
    }
  }

  /// 清除错误
  @action
  void clearError() {
    errorMessage = null;
  }

  /// 清除当前消息
  @action
  void clearCurrentMessage() {
    currentMessage = null;
    countdownSeconds = null;
    _countdownTimer?.cancel();
  }

  /// 重置状态
  @action
  void reset() {
    messages.clear();
    sentMessages.clear();
    receivedMessages.clear();
    screenshotDetectedMessages.clear();
    currentMessage = null;
    config = null;
    unreadCount = 0;
    errorMessage = null;
    countdownSeconds = null;
    _countdownTimer?.cancel();
    _countdownTimer = null;
  }

  /// 销毁
  void dispose() {
    _countdownTimer?.cancel();
  }
}

File: im-desktop/src/stores/self-destruct-message-store.ts
/**
 * 阅后即焚消息状态管理
 * 
 * @author IM Development Team
 * @since 1.0.0
 */

import { makeAutoObservable, runInAction } from 'mobx';
import {
  SelfDestructMessage,
  SelfDestructConfig,
  CreateSelfDestructMessageRequest,
  ReadSelfDestructMessageResponse,
  ScreenshotDetectResponse,
} from '../types/self-destruct-message';
import { selfDestructMessageApi } from '../api/self-destruct-message-api';

export class SelfDestructMessageStore {
  messages: SelfDestructMessage[] = [];
  sentMessages: SelfDestructMessage[] = [];
  receivedMessages: SelfDestructMessage[] = [];
  screenshotDetectedMessages: SelfDestructMessage[] = [];
  currentMessage: SelfDestructMessage | null = null;
  config: SelfDestructConfig | null = null;
  unreadCount: number = 0;
  isLoading: boolean = false;
  errorMessage: string | null = null;
  isReading: boolean = false;
  countdownSeconds: number | null = null;
  private countdownInterval: NodeJS.Timeout | null = null;

  constructor() {
    makeAutoObservable(this);
  }

  get activeMessages(): SelfDestructMessage[] {
    return this.messages.filter(m => !m.isDestroyed);
  }

  get destroyedMessages(): SelfDestructMessage[] {
    return this.messages.filter(m => m.isDestroyed);
  }

  get hasUnread(): boolean {
    return this.unreadCount > 0;
  }

  get hasScreenshotDetected(): boolean {
    return this.screenshotDetectedMessages.length > 0;
  }

  async fetchConfig(): Promise<void> {
    runInAction(() => {
      this.isLoading = true;
      this.errorMessage = null;
    });

    try {
      const config = await selfDestructMessageApi.getConfig();
      runInAction(() => {
        this.config = config;
      });
    } catch (error) {
      runInAction(() => {
        this.errorMessage = `获取配置失败: ${error}`;
      });
    } finally {
      runInAction(() => {
        this.isLoading = false;
      });
    }
  }

  async fetchMessages(conversationId: string, page: number = 0, size: number = 20): Promise<void> {
    runInAction(() => {
      this.isLoading = true;
      this.errorMessage = null;
    });

    try {
      const messages = await selfDestructMessageApi.getMessagesByConversation(conversationId, page, size);
      runInAction(() => {
        if (page === 0) {
          this.messages = messages;
        } else {
          this.messages.push(...messages);
        }
      });
    } catch (error) {
      runInAction(() => {
        this.errorMessage = `获取消息失败: ${error}`;
      });
    } finally {
      runInAction(() => {
        this.isLoading = false;
      });
    }
  }

  async fetchSentMessages(): Promise<void> {
    runInAction(() => {
      this.isLoading = true;
      this.errorMessage = null;
    });

    try {
      const messages = await selfDestructMessageApi.getSentMessages();
      runInAction(() => {
        this.sentMessages = messages;
      });
    } catch (error) {
      runInAction(() => {
        this.errorMessage = `获取发送消息失败: ${error}`;
      });
    } finally {
      runInAction(() => {
        this.isLoading = false;
      });
    }
  }

  async fetchReceivedMessages(): Promise<void> {
    runInAction(() => {
      this.isLoading = true;
      this.errorMessage = null;
    });

    try {
      const messages = await selfDestructMessageApi.getReceivedMessages();
      runInAction(() => {
        this.receivedMessages = messages;
      });
    } catch (error) {
      runInAction(() => {
        this.errorMessage = `获取接收消息失败: ${error}`;
      });
    } finally {
      runInAction(() => {
        this.isLoading = false;
      });
    }
  }

  async fetchUnreadCount(): Promise<void> {
    try {
      const { count } = await selfDestructMessageApi.getUnreadCount();
      runInAction(() => {
        this.unreadCount = count;
      });
    } catch (error) {
      runInAction(() => {
        this.errorMessage = `获取未读数量失败: ${error}`;
      });
    }
  }

  async fetchUnreadCountByConversation(conversationId: string): Promise<number> {
    try {
      const { count } = await selfDestructMessageApi.getUnreadCountByConversation(conversationId);
      return count;
    } catch (error) {
      runInAction(() => {
        this.errorMessage = `获取未读数量失败: ${error}`;
      });
      return 0;
    }
  }

  async createMessage(request: CreateSelfDestructMessageRequest): Promise<SelfDestructMessage | null> {
    runInAction(() => {
      this.isLoading = true;
      this.errorMessage = null;
    });

    try {
      const message = await selfDestructMessageApi.createMessage(request);
      runInAction(() => {
        this.messages.unshift(message);
        this.sentMessages.unshift(message);
      });
      return message;
    } catch (error) {
      runInAction(() => {
        this.errorMessage = `创建消息失败: ${error}`;
      });
      return null;
    } finally {
      runInAction(() => {
        this.isLoading = false;
      });
    }
  }

  async readMessage(messageId: string): Promise<ReadSelfDestructMessageResponse | null> {
    runInAction(() => {
      this.isReading = true;
      this.errorMessage = null;
    });

    try {
      const response = await selfDestructMessageApi.readMessage(messageId);
      
      runInAction(() => {
        const index = this.messages.findIndex(m => m.id === messageId);
        if (index !== -1) {
          this.messages[index] = {
            ...this.messages[index],
            isRead: true,
            remainingSeconds: response.remainingSeconds,
          };
        }

        this.countdownSeconds = response.remainingSeconds;
        this.startCountdown(messageId);
      });

      return response;
    } catch (error) {
      runInAction(() => {
        this.errorMessage = `阅读消息失败: ${error}`;
      });
      return null;
    } finally {
      runInAction(() => {
        this.isReading = false;
      });
    }
  }

  private startCountdown(messageId: string): void {
    if (this.countdownInterval) {
      clearInterval(this.countdownInterval);
    }

    this.countdownInterval = setInterval(() => {
      runInAction(() => {
        if (this.countdownSeconds === null || this.countdownSeconds <= 0) {
          this.destroyMessageInternal(messageId);
          if (this.countdownInterval) {
            clearInterval(this.countdownInterval);
            this.countdownInterval = null;
          }
        } else {
          this.countdownSeconds--;
          const index = this.messages.findIndex(m => m.id === messageId);
          if (index !== -1) {
            this.messages[index].remainingSeconds = this.countdownSeconds;
          }
        }
      });
    }, 1000);
  }

  private destroyMessageInternal(messageId: string): void {
    const index = this.messages.findIndex(m => m.id === messageId);
    if (index !== -1) {
      this.messages[index] = {
        ...this.messages[index],
        isDestroyed: true,
        messageContent: undefined,
        remainingSeconds: 0,
      };
    }

    selfDestructMessageApi.destroyMessage(messageId).catch(() => {
      // 忽略错误
    });
  }

  async destroyMessage(messageId: string): Promise<boolean> {
    try {
      await selfDestructMessageApi.destroyMessage(messageId);
      
      runInAction(() => {
        const index = this.messages.findIndex(m => m.id === messageId);
        if (index !== -1) {
          this.messages[index] = {
            ...this.messages[index],
            isDestroyed: true,
            messageContent: undefined,
            remainingSeconds: 0,
          };
        }

        if (this.countdownInterval) {
          clearInterval(this.countdownInterval);
          this.countdownInterval = null;
        }
      });

      return true;
    } catch (error) {
      runInAction(() => {
        this.errorMessage = `销毁消息失败: ${error}`;
      });
      return false;
    }
  }

  async deleteMessage(messageId: string): Promise<boolean> {
    try {
      await selfDestructMessageApi.deleteMessage(messageId);
      
      runInAction(() => {
        this.messages = this.messages.filter(m => m.id !== messageId);
        this.sentMessages = this.sentMessages.filter(m => m.id !== messageId);
      });

      return true;
    } catch (error) {
      runInAction(() => {
        this.errorMessage = `删除消息失败: ${error}`;
      });
      return false;
    }
  }

  async detectScreenshot(messageId: string): Promise<ScreenshotDetectResponse | null> {
    try {
      const response = await selfDestructMessageApi.detectScreenshot(messageId);
      
      runInAction(() => {
        const index = this.messages.findIndex(m => m.id === messageId);
        if (index !== -1) {
          this.messages[index] = {
            ...this.messages[index],
            screenshotDetected: true,
            screenshotCount: response.totalCount,
          };
        }
      });

      return response;
    } catch (error) {
      runInAction(() => {
        this.errorMessage = `截图检测失败: ${error}`;
      });
      return null;
    }
  }

  async fetchScreenshotDetectedMessages(): Promise<void> {
    runInAction(() => {
      this.isLoading = true;
      this.errorMessage = null;
    });

    try {
      const messages = await selfDestructMessageApi.getScreenshotDetectedMessages();
      runInAction(() => {
        this.screenshotDetectedMessages = messages;
      });
    } catch (error) {
      runInAction(() => {
        this.errorMessage = `获取截图检测消息失败: ${error}`;
      });
    } finally {
      runInAction(() => {
        this.isLoading = false;
      });
    }
  }

  addNewMessage(message: SelfDestructMessage): void {
    if (!this.messages.some(m => m.id === message.id)) {
      this.messages.unshift(message);
      this.receivedMessages.unshift(message);
      this.unreadCount++;
    }
  }

  updateMessage(message: SelfDestructMessage): void {
    const index = this.messages.findIndex(m => m.id === message.id);
    if (index !== -1) {
      this.messages[index] = message;
    }
  }

  clearError(): void {
    this.errorMessage = null;
  }

  clearCurrentMessage(): void {
    this.currentMessage = null;
    this.countdownSeconds = null;
    if (this.countdownInterval) {
      clearInterval(this.countdownInterval);
      this.countdownInterval = null;
    }
  }

  reset(): void {
    this.messages = [];
    this.sentMessages = [];
    this.receivedMessages = [];
    this.screenshotDetectedMessages = [];
    this.currentMessage = null;
    this.config = null;
    this.unreadCount = 0;
    this.errorMessage = null;
    this.countdownSeconds = null;
    if (this.countdownInterval) {
      clearInterval(this.countdownInterval);
      this.countdownInterval = null;
    }
  }
}

export const selfDestructMessageStore = new SelfDestructMessageStore();
